{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single-cell RNA-seq analysis of HGSOC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/commons/home/jameslee/miniconda3/envs/ml_env/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/gpfs/commons/home/jameslee/miniconda3/envs/ml_env/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/gpfs/commons/home/jameslee/miniconda3/envs/ml_env/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/gpfs/commons/home/jameslee/miniconda3/envs/ml_env/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/gpfs/commons/home/jameslee/miniconda3/envs/ml_env/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/gpfs/commons/home/jameslee/miniconda3/envs/ml_env/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "Matplotlib requires numpy>=1.23; you have 1.22.3",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mnumpy\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mscanpy\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01msc\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpyplot\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mplt\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mcollections\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m defaultdict\n",
      "File \u001b[0;32m~/miniconda3/envs/ml_env/lib/python3.10/site-packages/scanpy/__init__.py:24\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[38;5;66;03m# the actual API\u001b[39;00m\n\u001b[1;32m     22\u001b[0m \u001b[38;5;66;03m# (start with settings as several tools are using it)\u001b[39;00m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_settings\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m settings, Verbosity\n\u001b[0;32m---> 24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m tools \u001b[38;5;28;01mas\u001b[39;00m tl\n\u001b[1;32m     25\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m preprocessing \u001b[38;5;28;01mas\u001b[39;00m pp\n\u001b[1;32m     26\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m plotting \u001b[38;5;28;01mas\u001b[39;00m pl\n",
      "File \u001b[0;32m~/miniconda3/envs/ml_env/lib/python3.10/site-packages/scanpy/tools/__init__.py:17\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_leiden\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m leiden\n\u001b[1;32m     16\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_louvain\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m louvain\n\u001b[0;32m---> 17\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_sim\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m sim\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_score_genes\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m score_genes, score_genes_cell_cycle\n\u001b[1;32m     19\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_dendrogram\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m dendrogram\n",
      "File \u001b[0;32m~/miniconda3/envs/ml_env/lib/python3.10/site-packages/scanpy/tools/_sim.py:23\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mscipy\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01msp\u001b[39;00m\n\u001b[1;32m     21\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01manndata\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m AnnData\n\u001b[0;32m---> 23\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m _utils, readwrite, logging \u001b[38;5;28;01mas\u001b[39;00m logg\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_settings\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m settings\n\u001b[1;32m     27\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21msim\u001b[39m(\n\u001b[1;32m     28\u001b[0m     model: Literal[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mkrumsiek11\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtoggleswitch\u001b[39m\u001b[38;5;124m'\u001b[39m],\n\u001b[1;32m     29\u001b[0m     params_file: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     37\u001b[0m     writedir: Optional[Union[\u001b[38;5;28mstr\u001b[39m, Path]] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m     38\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m AnnData:\n",
      "File \u001b[0;32m~/miniconda3/envs/ml_env/lib/python3.10/site-packages/scanpy/readwrite.py:24\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01manndata\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[1;32m     15\u001b[0m     AnnData,\n\u001b[1;32m     16\u001b[0m     read_csv,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     21\u001b[0m     read_text,\n\u001b[1;32m     22\u001b[0m )\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01manndata\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m read \u001b[38;5;28;01mas\u001b[39;00m read_h5ad\n\u001b[0;32m---> 24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mimage\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m imread\n\u001b[1;32m     26\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m logging \u001b[38;5;28;01mas\u001b[39;00m logg\n\u001b[1;32m     27\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_settings\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m settings\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/matplotlib/__init__.py:264\u001b[0m\n\u001b[1;32m    259\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m parse_version(module\u001b[38;5;241m.\u001b[39m__version__) \u001b[38;5;241m<\u001b[39m parse_version(minver):\n\u001b[1;32m    260\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMatplotlib requires \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmodname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m>=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mminver\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m; \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    261\u001b[0m                               \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myou have \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmodule\u001b[38;5;241m.\u001b[39m__version__\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 264\u001b[0m \u001b[43m_check_versions\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    267\u001b[0m \u001b[38;5;66;03m# The decorator ensures this always returns the same handler (and it is only\u001b[39;00m\n\u001b[1;32m    268\u001b[0m \u001b[38;5;66;03m# attached once).\u001b[39;00m\n\u001b[1;32m    269\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mcache\n\u001b[1;32m    270\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_ensure_handler\u001b[39m():\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/matplotlib/__init__.py:260\u001b[0m, in \u001b[0;36m_check_versions\u001b[0;34m()\u001b[0m\n\u001b[1;32m    258\u001b[0m module \u001b[38;5;241m=\u001b[39m importlib\u001b[38;5;241m.\u001b[39mimport_module(modname)\n\u001b[1;32m    259\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m parse_version(module\u001b[38;5;241m.\u001b[39m__version__) \u001b[38;5;241m<\u001b[39m parse_version(minver):\n\u001b[0;32m--> 260\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMatplotlib requires \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmodname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m>=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mminver\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m; \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    261\u001b[0m                       \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myou have \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmodule\u001b[38;5;241m.\u001b[39m__version__\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mImportError\u001b[0m: Matplotlib requires numpy>=1.23; you have 1.22.3"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set random seed for reproducibility\n",
    "np.random.seed(42)\n",
    "\n",
    "# Set plotting style\n",
    "sc.settings.set_figure_params(dpi=100, frameon=False)\n",
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "plt.rcParams['figure.figsize'] = (8, 8)\n",
    "\n",
    "# Paths\n",
    "METADATA_PATH = \"/gpfs/commons/home/jameslee/HGSOC/metadata/GSE184880_metadata.csv\"\n",
    "CELLRANGER_DIR = \"/gpfs/commons/home/jameslee/HGSOC/output/cellranger\"\n",
    "OUTPUT_DIR = \"/gpfs/commons/home/jameslee/HGSOC/output/scanpy\"\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Load metadata\n",
    "metadata = pd.read_csv(METADATA_PATH)\n",
    "print(f\"Loaded metadata with {len(metadata)} entries\")\n",
    "\n",
    "# Group SRR runs by GSM sample name\n",
    "sample_groups = defaultdict(list)\n",
    "for _, row in metadata.iterrows():\n",
    "    sample_groups[row['Sample Name']].append(row['Run'])\n",
    "\n",
    "print(f\"Found {len(sample_groups)} unique GSM samples\")\n",
    "\n",
    "# Process each GSM sample\n",
    "all_samples = []\n",
    "\n",
    "for gsm_name, srr_list in sample_groups.items():\n",
    "    print(f\"\\nProcessing {gsm_name} with {len(srr_list)} sequencing runs\")\n",
    "    \n",
    "    # Check if all SRR data exists\n",
    "    srr_adatas = []\n",
    "    for srr_id in srr_list:\n",
    "        h5_path = os.path.join(CELLRANGER_DIR, srr_id, \"outs\", \"filtered_feature_bc_matrix.h5\")\n",
    "        if not os.path.exists(h5_path):\n",
    "            print(f\"Warning: CellRanger output for {srr_id} not found at {h5_path}\")\n",
    "            continue\n",
    "        \n",
    "        # Load data\n",
    "        print(f\"Loading {srr_id}...\")\n",
    "        adata = sc.read_10x_h5(h5_path)\n",
    "        \n",
    "        # Add metadata\n",
    "        meta_row = metadata[metadata['Run'] == srr_id].iloc[0]\n",
    "        adata.obs['run_id'] = srr_id\n",
    "        adata.obs['sample_name'] = gsm_name\n",
    "        adata.obs['age'] = meta_row['age']\n",
    "        adata.obs['pathology'] = meta_row['pathology']\n",
    "        adata.obs['tissue_type'] = meta_row['tissue_type']\n",
    "        adata.obs['tumor_stage'] = meta_row['tumor_stage']\n",
    "        \n",
    "        # Add a unique barcode suffix to avoid conflicts when merging\n",
    "        adata.obs_names = [f\"{bc}_{srr_id}\" for bc in adata.obs_names]\n",
    "        \n",
    "        # Basic QC metrics\n",
    "        sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "        \n",
    "        srr_adatas.append(adata)\n",
    "    \n",
    "    if not srr_adatas:\n",
    "        print(f\"No data found for {gsm_name}, skipping\")\n",
    "        continue\n",
    "    \n",
    "    # Merge all sequencing runs for this sample\n",
    "    if len(srr_adatas) > 1:\n",
    "        combined = srr_adatas[0].concatenate(srr_adatas[1:], batch_key='sequencing_run')\n",
    "        print(f\"Merged {len(srr_adatas)} runs for {gsm_name}, resulting in {combined.n_obs} cells\")\n",
    "    else:\n",
    "        combined = srr_adatas[0]\n",
    "        combined.obs['sequencing_run'] = '0'\n",
    "        print(f\"Single run for {gsm_name}, {combined.n_obs} cells\")\n",
    "    \n",
    "    # Basic preprocessing for this sample\n",
    "    print(f\"Preprocessing {gsm_name}...\")\n",
    "    \n",
    "    # Filter cells based on QC metrics (adjust thresholds as needed)\n",
    "    sc.pp.filter_cells(combined, min_genes=200)\n",
    "    sc.pp.filter_genes(combined, min_cells=3)\n",
    "    \n",
    "    # Filter out cells with high mitochondrial gene fraction\n",
    "    combined.var['mt'] = combined.var_names.str.startswith('MT-')\n",
    "    sc.pp.calculate_qc_metrics(combined, qc_vars=['mt'], inplace=True)\n",
    "    combined = combined[combined.obs['pct_counts_mt'] < 20].copy()\n",
    "    \n",
    "    print(f\"After QC: {combined.n_obs} cells, {combined.n_vars} genes\")\n",
    "    \n",
    "    # Save the processed individual sample\n",
    "    combined.write(os.path.join(OUTPUT_DIR, f\"{gsm_name}_processed.h5ad\"))\n",
    "    \n",
    "    # Store for integration later\n",
    "    all_samples.append(combined)\n",
    "\n",
    "# Integrate all samples\n",
    "if len(all_samples) > 0:\n",
    "    print(\"\\nIntegrating all samples...\")\n",
    "    integrated = all_samples[0].concatenate(all_samples[1:], batch_key='sample_name')\n",
    "    print(f\"Integrated dataset: {integrated.n_obs} cells, {integrated.n_vars} genes\")\n",
    "    \n",
    "    # Normalize and log transform\n",
    "    sc.pp.normalize_total(integrated, target_sum=1e4)\n",
    "    sc.pp.log1p(integrated)\n",
    "    \n",
    "    # Find variable genes\n",
    "    sc.pp.highly_variable_genes(integrated, min_mean=0.0125, max_mean=3, min_disp=0.5, batch_key='sample_name')\n",
    "    print(f\"Found {sum(integrated.var.highly_variable)} highly variable genes\")\n",
    "    \n",
    "    # Scale data\n",
    "    sc.pp.scale(integrated, max_value=10)\n",
    "    \n",
    "    # Run PCA\n",
    "    sc.tl.pca(integrated, svd_solver='arpack')\n",
    "    \n",
    "    # Batch correction using harmony (if installed) or bbknn\n",
    "    try:\n",
    "        import harmonypy\n",
    "        sc.external.pp.harmony_integrate(integrated, 'sample_name')\n",
    "        use_rep = 'X_pca_harmony'\n",
    "    except ImportError:\n",
    "        print(\"Harmony not available, using standard PCA\")\n",
    "        use_rep = 'X_pca'\n",
    "    \n",
    "    # Compute neighborhood graph\n",
    "    sc.pp.neighbors(integrated, n_neighbors=10, n_pcs=30, use_rep=use_rep)\n",
    "    \n",
    "    # Run UMAP and clustering\n",
    "    sc.tl.umap(integrated)\n",
    "    sc.tl.leiden(integrated, resolution=0.5)\n",
    "    \n",
    "    # Calculate marker genes\n",
    "    sc.tl.rank_genes_groups(integrated, 'leiden', method='wilcoxon')\n",
    "    \n",
    "    # Plot results\n",
    "    sc.pl.umap(integrated, color=['leiden', 'sample_name', 'tissue_type', 'tumor_stage'], \n",
    "               ncols=2, save=f\"_umap_clusters_samples.png\")\n",
    "    \n",
    "    # Create separate UMAPs for specific groupings\n",
    "    sc.pl.umap(integrated, color='tissue_type', save=\"_umap_tissue_type.png\")\n",
    "    sc.pl.umap(integrated, color='tumor_stage', save=\"_umap_tumor_stage.png\")\n",
    "    \n",
    "    # Save the integrated dataset\n",
    "    integrated.write(os.path.join(OUTPUT_DIR, \"integrated_dataset.h5ad\"))\n",
    "    \n",
    "    # Save marker genes to CSV\n",
    "    result = integrated.uns['rank_genes_groups']\n",
    "    groups = result['names'].dtype.names\n",
    "    \n",
    "    marker_df = pd.DataFrame({\n",
    "        group + '_' + key: result[key][group]\n",
    "        for group in groups\n",
    "        for key in ['names', 'pvals', 'pvals_adj', 'scores']\n",
    "    })\n",
    "    \n",
    "    marker_df.to_csv(os.path.join(OUTPUT_DIR, \"marker_genes.csv\"))\n",
    "    \n",
    "    print(f\"Analysis complete. Results saved to {OUTPUT_DIR}\")\n",
    "else:\n",
    "    print(\"No samples were processed successfully!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
